src/frontend/src/optimizer/rule/grouping_sets_to_expand_rule.rs

```diff
@@ -85,35 +87,92 @@ impl Rule for GroupingSetsToExpandRule {
             .iter()
             .map(|set| set.indices().collect_vec())
             .collect_vec();
-        let expand = LogicalExpand::create(input, column_subset);
+
+        let expand = LogicalExpand::create(input, column_subset.clone());
         // Add the expand flag.
         group_keys.extend(std::iter::once(expand.schema().len() - 1));
 
         let mut input_col_change =
             ColIndexMapping::with_shift_offset(input_schema_len, input_schema_len as isize);
 
-        // Shift agg_call to the original input columns
-        let new_agg_calls = agg_calls
-            .iter()
-            .cloned()
-            .map(|mut agg_call| {
+        // Grouping agg calls need to be transformed into a project expression, and other agg calls
+        // need to shift their `input_ref`.
+        let mut project_exprs = vec![];
+        let mut new_agg_calls = vec![];
+        for mut agg_call in agg_calls {
+            // Deal with grouping agg call for grouping sets.
+            if agg_call.agg_kind == AggKind::Grouping {
+                let mut grouping_values = vec![];
+                let args = agg_call
+                    .inputs
+                    .iter()
+                    .map(|input_ref| input_ref.index)
+                    .collect_vec();
+                for subset in &column_subset {
+                    let mut value = 0;
+                    for arg in &args {
+                        value <<= 1;
+                        if !subset.contains(arg) {
+                            value += 1;
+                        }
+                    }
+                    grouping_values.push(value);
+                }
+
+                let mut case_inputs = vec![];
+                for (i, grouping_value) in grouping_values.into_iter().enumerate() {
+                    let condition = ExprImpl::FunctionCall(
+                        FunctionCall::new_unchecked(
+                            ExprType::Equal,
+                            vec![
+                                ExprImpl::literal_bigint(i as i64),
+                                ExprImpl::InputRef(
+                                    InputRef::new(original_group_keys_num, DataType::Int64).into(),
```

Comment:
Would you mind changing `original_group_keys_num` to something like `flag_col_idx` which will looks clearer?